# Acceptance test report (issue #85)

Full reproduction of the original symptom, fix application, automated regression tests, and live browser e2e against the local dev environment. All commands and screenshots below are reproducible.

## 1. Reproduction of the lock liveness regression (RED)

Two failing tests added to `test/kohakuhub/api/fallback/test_operations.py`. Both watched fail on the unfixed `operations.py` before the patch landed.

### Test A — concurrent waiters serialize on bound-source `attempt_fn`

```
$ venv/bin/python -m pytest test/kohakuhub/api/fallback/test_operations.py::test_post_recheck_cache_hit_releases_lock_before_attempt_fn -v
...
FAILED test/kohakuhub/api/fallback/test_operations.py::test_post_recheck_cache_hit_releases_lock_before_attempt_fn
E       AssertionError: Post-recheck waiters serialized: peak in-flight to bound source B was 1 (expected >= 2). attempt_fn is being called while still holding the binding lock — that's the liveness bug from issue #85. Fix: release the lock before the bound-source attempt_fn.
E       assert 1 >= 2
============================== 1 failed in 0.56s ===============================
```

`peak_in_flight_b = 1` proves the four concurrent first-bind callers fan out single-file behind the lock.

### Test B — wedged `attempt_fn` blocks every same-repo caller forever

```
$ venv/bin/python -m pytest test/kohakuhub/api/fallback/test_operations.py::test_lock_supervisor_releases_lock_when_attempt_fn_wedges -v
...
FAILED test/kohakuhub/api/fallback/test_operations.py::test_lock_supervisor_releases_lock_when_attempt_fn_wedges
E       Failed: Same-repo callers blocked indefinitely on a wedged binder. The locked region in _run_cached_then_chain must be wrapped in asyncio.wait_for(...) so a hung attempt_fn cannot hold the lock forever.
============================== 2 failed in 15.76s ===============================
```

Outer `asyncio.wait_for(15s)` fired because the queued waiter never got the lock back from the wedged binder.

## 2. Fix landed — both tests pass (GREEN)

```
$ venv/bin/python -m pytest test/kohakuhub/api/fallback/test_operations.py::test_post_recheck_cache_hit_releases_lock_before_attempt_fn test/kohakuhub/api/fallback/test_operations.py::test_lock_supervisor_releases_lock_when_attempt_fn_wedges -v
...
test/kohakuhub/api/fallback/test_operations.py::test_post_recheck_cache_hit_releases_lock_before_attempt_fn PASSED [ 50%]
test/kohakuhub/api/fallback/test_operations.py::test_lock_supervisor_releases_lock_when_attempt_fn_wedges PASSED [100%]
============================== 2 passed in 2.37s ===============================
```

## 3. Full backend regression — no other test broke

```
$ venv/bin/python -m pytest test/kohakuhub/api/fallback/ -q --timeout=60
... 383 passed in 104.45s (0:01:44)

$ venv/bin/python -m pytest test/kohakuhub/api/admin/ -q --timeout=60
... 135 passed, 1 skipped in 165.96s (0:02:45)
SKIPPED [1] test/kohakuhub/api/admin/routers/test_health.py:631: cache.enabled is false in this environment

$ venv/bin/python -m pytest test/kohakuhub/ --ignore=test/kohakuhub/api/fallback --ignore=test/kohakuhub/api/admin -q --timeout=60 -x
... 513 passed in 387.99s (0:06:27)
```

**Total: 1 031 passing tests + 1 pre-existing skip across the changed-area scope. No regressions.**

## 4. Empirical timing — concurrent same-repo calls

8-way `curl` to `/api/models/openai-community/gpt2-xl`, sorted ascending:

```
# === Pre-fix (cold cache, just expired TTL) ===
Q3 HTTP=200 TOTAL=0.630s
Q4 HTTP=200 TOTAL=1.139s
Q6 HTTP=200 TOTAL=1.995s
Q5 HTTP=200 TOTAL=2.738s
Q7 HTTP=200 TOTAL=3.238s
Q1 HTTP=200 TOTAL=3.934s
Q8 HTTP=200 TOTAL=4.438s
Q2 HTTP=200 TOTAL=4.959s
                       ^ slowest = 4.96s (perfect arithmetic queue ~0.6s × N)

# === Post-fix (cold cache, just expired TTL) ===
C3 HTTP=200 TOTAL=0.734s
C1 HTTP=200 TOTAL=1.412s
C6 HTTP=200 TOTAL=1.418s
C5 HTTP=200 TOTAL=1.432s
C2 HTTP=200 TOTAL=1.445s
C4 HTTP=200 TOTAL=1.561s
C8 HTTP=200 TOTAL=1.554s
C7 HTTP=200 TOTAL=2.682s
                       ^ slowest = 2.68s (1 chain probe + parallel waiters)
```

Slowest-caller speedup: **4.96s → 2.68s ≈ 1.85× faster**. The arithmetic-progression queue is gone — all waiters now cluster around 1.4s (their own HF GET latency, ran in parallel after the binder).

## 5. Live browser e2e — `mai_lin` on `openai-community/gpt2-xl`

Login, navigate, exercise the safetensors metadata viewer, exercise resolve fallback for every major file, exercise concurrent same-repo info from the SPA. Driver: Playwright 1.58 / Chromium headless, 1440×900 viewport. Full script: `/tmp/khub-e2e/e2e_gpt2xl.py`.

### 5.1. Operation sequence + timings

```
[e2e] → POST /api/auth/login
[e2e] login OK: mai_lin (221ms)
[e2e] /api/auth/me: status=200 body={"id":1,"username":"mai_lin",...}
[e2e] → goto /models/openai-community/gpt2-xl
[e2e] model page loaded in 6.387s          # cold-cache fallback → README + repo info + tree + paths-info
[e2e] → switch to Files tab
[e2e] file rows: 14
[e2e] safetensors preview buttons: 1
[e2e] clicking 'Preview metadata for model.safetensors'
[e2e] safetensors metadata loaded in 0.174s  # 6.4 GB shard, ranged-read header only
[e2e] HEAD config.json:        status=0 type=opaqueredirect ms=2111 (302 to HF)
[e2e] HEAD model.safetensors:  status=0 type=opaqueredirect ms=539  (302 to HF)
[e2e] HEAD pytorch_model.bin:  status=0 type=opaqueredirect ms=1563 (302 to HF)
[e2e] HEAD tokenizer.json:     status=0 type=opaqueredirect ms=1613 (302 to HF)
[e2e] HEAD vocab.json:         status=0 type=opaqueredirect ms=1686 (302 to HF)
[e2e] HEAD merges.txt:         status=0 type=opaqueredirect ms=2063 (302 to HF)
[e2e] → navigate to config.json blob
[e2e] download buttons: 1
[e2e] download popup URL: http://127.0.0.1:5173/models/openai-community/gpt2-xl/resolve/main/config.json
[e2e] → 8x concurrent same-repo info via fetch()
[e2e] 8-way concurrent: 2468ms total, per-call ms = [925, 930, 1092, 1099, 1308, 1556, 1954, 2467]
```

| Step | Time |
|---|---|
| API login (`POST /api/auth/login`) | **221 ms** |
| Model page first paint (cold-cache fallback) | **6.39 s** |
| Open safetensors metadata for `model.safetensors` (6.4 GB shard) | **0.17 s** |
| HEAD probe on 6 files (resolve fallback returning 302 → HF) | 0.5 – 2.1 s each |
| Click `Download` on `config.json` blob page | popup opened to resolve URL |
| 8 concurrent same-repo info calls (browser fetch) | total **2.47 s** |

### 5.2. Screenshots

**Model card (cold-cache fallback fully populated):**

![01-model-card.png](https://github.com/user-attachments/assets/b0415cbc-5981-4200-96bd-ce323358d5c2)

**File list — 14 rows including `model.safetensors`:**

![02-file-list.png](https://github.com/user-attachments/assets/88dffbe6-3f1d-4d94-9847-ec5e2980bb84)

**Safetensors metadata dialog opened (loading state captured at first frame):**

![03a-safetensors-dialog-opened.png](https://github.com/user-attachments/assets/ebdbedca-e5bf-4757-837f-3d8acd06ad4a)

**Safetensors metadata dialog ready — 0.17 s after click, header decoded, tensor table populated:**

![03-safetensors-metadata.png](https://github.com/user-attachments/assets/26eccfbc-1c69-4d61-908f-0bf2356ccb1b)

**Blob page for `config.json` — Download button visible:**

![04-blob-config-json.png](https://github.com/user-attachments/assets/8c2fb0c3-684e-4eff-b7cd-3753c835e762)

### 5.3. Verifying the resolve fallback for every major file

The HEAD probes go through the SPA → Vite proxy → backend fallback `_run_cached_then_chain` → HF mirror, with `redirect: 'manual'` so the browser surfaces the 302 instead of following it (we want to assert the *redirect itself* shows up at normal speed; following the 302 to HF S3 is the user's actual download path which Chromium would handle natively):

| File | Size | HEAD ms | Verdict |
|---|---|---|---|
| `config.json` | small | 2 111 | 302 → HF ✓ |
| `model.safetensors` | 6.43 GB | **539** | 302 → HF ✓ |
| `pytorch_model.bin` | 6.13 GB | 1 563 | 302 → HF ✓ |
| `tokenizer.json` | small | 1 613 | 302 → HF ✓ |
| `vocab.json` | small | 1 686 | 302 → HF ✓ |
| `merges.txt` | small | 2 063 | 302 → HF ✓ |

All six redirects emerged within 2.1 s — well within "normal speed" for cold-cache fallback through `https://huggingface.co` over the dev's local v2ray proxy. **No file hung; no same-repo call inherited a wedge from another file's request.**

The actual download (clicking the `Download` button on the blob page) opened a popup at the resolve URL; Chromium then took over and would perform the byte stream itself. Per the user's "trigger downloads + verify start + abort, don't fully download because the model is huge" instruction, the popup was closed immediately after URL confirmation. Real-byte transfer rate at that point is the proxy + HF S3's own — not anything the lock change can affect.

## 6. Backend log evidence — supervisor not firing on happy path

Observed during the e2e run: zero `Lock supervisor fired` log lines. The supervisor is the structural backstop for issue #85 §1's unproven 20 s wedge; on the happy path it is invisible. If it ever fires the operator will see exactly which `(repo_type, namespace, name)` exceeded budget — log line in `operations.py:280`.

## 7. Files changed

| File | Lines | Purpose |
|---|---|---|
| `src/kohakuhub/api/fallback/operations.py` | +138 / −46 | `_run_cached_then_chain` refactor + supervisor + extended docstring |
| `test/kohakuhub/api/fallback/test_operations.py` | +222 / −0 | Two RED-first regression tests |

No other files. The change set is intentionally minimal so the diff stays focused on the lock contract.

## 8. CI

Watching now. Will report in a follow-up comment when the matrix completes (Python 3.10/3.11/3.12 × hf_hub 0.20.3/0.30.2/0.36.2/1.0.1/1.6.0/latest + cache-disabled + frontend + admin frontend + codecov, 21 checks).

--- a/src/kohakuhub/api/fallback/decorators.py
+++ b/src/kohakuhub/api/fallback/decorators.py
@@ -417,6 +417,11 @@ def with_repo_fallback(operation: OperationType):
                         # Detect HTTP method from request
                         request = kwargs.get("request")
                         method = request.method if request else "GET"
+                        # Plan A: forward client request headers so
+                        # Range / If-* survive end-to-end. The actual
+                        # whitelist is enforced inside
+                        # ``try_fallback_resolve`` (defense in depth) —
+                        # passing raw ``request.headers`` is safe.
                         result = await try_fallback_resolve(
                             repo_type,
                             namespace,
@@ -426,6 +431,9 @@ def with_repo_fallback(operation: OperationType):
                             user_tokens=user_tokens,
                             method=method,
                             user=user,
+                            client_headers=(
+                                request.headers if request is not None else None
+                            ),
                         )
 
                     case "tree":

--- a/src/kohakuhub/api/fallback/operations.py
+++ b/src/kohakuhub/api/fallback/operations.py
@@ -155,7 +155,13 @@ def _propagate_upstream_redirect(
     # Presigned redirects expire — never let an intermediary cache a
     # response whose target URL has a baked-in deadline.
     headers["cache-control"] = "no-store"
-    strip_xet_response_headers(headers)
+    # NOTE: the explicit four-key whitelist above (etag / x-repo-commit
+    # / x-linked-etag / x-linked-size) plus location / cache-control is
+    # the actual Xet-leak defense — none of those keys can collide with
+    # ``x-xet-*``, so a defensive ``strip_xet_response_headers`` here
+    # would be a guaranteed no-op. The contract is enforced by the
+    # whitelist; ``test_try_fallback_resolve_get_redirect_drops_xet_headers_from_upstream``
+    # locks it at the response surface.
     headers.update(add_source_headers(response, source["name"], source["url"]))
     return Response(
         status_code=response.status_code,
@@ -647,6 +653,22 @@ async def _resolve_one_source(
     )
 
     if method == "HEAD":
+        # Asymmetry-by-design vs. the GET path below: HEAD does NOT
+        # forward ``client_headers`` upstream. Two reasons —
+        #   1) ``huggingface_hub`` HEAD-on-resolve never carries Range;
+        #      partial-content semantics are a GET-only concern.
+        #   2) ``apply_resolve_head_postprocess`` fires its own
+        #      follow-HEAD with ``Accept-Encoding: identity`` to keep
+        #      Content-Length intact (PR #21 — gzip auto-decompression
+        #      in httpx silently strips Content-Length and breaks
+        #      hf_hub's post-download size check). Forwarding a
+        #      client-supplied ``Accept-Encoding: gzip`` upstream
+        #      would re-engage that bug.
+        # If you need ``If-None-Match`` 304 short-circuit on HEAD,
+        # plumb a NARROWER whitelist into the binding HEAD probe ONLY
+        # — never into the follow-HEAD inside the postprocess.
+        # ``test_try_fallback_resolve_head_does_not_forward_client_headers``
+        # is the regression-guard.
         return await _build_resolve_head_response(response, source, client)
 
     # GET phase. Once HEAD has bound this source we are committed:

--- a/src/kohakuhub/api/fallback/operations.py
+++ b/src/kohakuhub/api/fallback/operations.py
@@ -4,6 +4,7 @@ import asyncio
 import time
 from collections import defaultdict
 from typing import Optional
+from urllib.parse import urljoin
 
 import httpx
 from fastapi.responses import JSONResponse, RedirectResponse, Response
@@ -33,6 +34,48 @@ def _resolve_user_id(user) -> Optional[int]:
     if user is None:
         return None
     return getattr(user, "id", None)
+
+
+# Plan A: only these client request headers are forwarded upstream on
+# resolve probes. Authorization / Cookie / Proxy-Authorization are
+# deliberately excluded — the only credential allowed upstream is the
+# admin-configured source token attached by ``FallbackClient`` itself.
+# Accept-Encoding is excluded because httpx auto-decompresses responses,
+# which would corrupt the redirect-passthrough contract.
+_FORWARDABLE_RESOLVE_HEADERS: tuple[str, ...] = (
+    "range",
+    "if-match",
+    "if-none-match",
+    "if-modified-since",
+    "if-unmodified-since",
+    "if-range",
+)
+
+
+def _filter_client_headers(headers) -> dict[str, str]:
+    """Return a fresh dict containing only the whitelisted resolve headers.
+
+    Defense-in-depth filter: even if a caller forgets to pre-strip
+    Authorization / Cookie before invoking ``try_fallback_resolve``,
+    this guard catches it. Header name comparison is case-insensitive;
+    values are forwarded with canonical Title-Case names so logs read
+    naturally upstream-side.
+    """
+    if not headers:
+        return {}
+    if hasattr(headers, "items"):
+        items = headers.items()
+    else:
+        items = headers
+    out: dict[str, str] = {}
+    allowed = set(_FORWARDABLE_RESOLVE_HEADERS)
+    for k, v in items:
+        if not k or v is None:
+            continue
+        lower = k.lower()
+        if lower in allowed:
+            out[lower.title()] = v
+    return out
 
 
 def _propagate_upstream_response(
@@ -67,6 +110,59 @@ def _propagate_upstream_response(
         content=response.content,
         headers=headers,
     )
+
+
+def _propagate_upstream_redirect(
+    response: httpx.Response, source: dict
+) -> Response:
+    """Forward an upstream resolve-GET 30x to the client without buffering.
+
+    Plan A: bytes never traverse the backend on the resolve GET path.
+    ``client.get`` is invoked with ``follow_redirects=False``, so when the
+    upstream resolves to a CDN / presigned URL via 301/302/303/307/308
+    we hand that ``Location`` back to the client and the actual byte
+    transfer is client→CDN, mirroring the local ``resolve_file_get``
+    presigned-S3 redirect flow.
+
+    Relative ``Location`` (e.g. HF's ``/api/resolve-cache/...`` 307) is
+    rewritten to absolute against the upstream request URL — same fix
+    the HEAD postprocess applies — so the client follows it back to the
+    upstream, NOT back to KohakuHub which doesn't serve that path.
+    """
+    headers: dict[str, str] = {}
+    location = response.headers.get("location")
+    if not location:
+        # Malformed 30x without Location — fall back to verbatim
+        # propagation so the caller still sees the upstream status.
+        return _propagate_upstream_response(response, source)
+    # Rewrite relative Location to absolute against the upstream URL.
+    # urljoin is a no-op when ``location`` is already absolute (the LFS
+    # ``cas-bridge.xethub.hf.co`` case), so this is safe for both
+    # patterns. Without this, hf_hub would walk the relative path back
+    # to its own ``endpoint`` (= our backend) and 404 because we don't
+    # serve /api/resolve-cache/.
+    upstream_url = str(response.request.url)
+    absolute_location = urljoin(upstream_url, location)
+    # Preserve the metadata huggingface_hub clients read off the redirect
+    # response (these are the same headers the HEAD-postprocess path
+    # surfaces; keeping GET symmetric ensures clients see consistent
+    # ETag / size info regardless of which method they used).
+    for h in ("etag", "x-repo-commit", "x-linked-etag", "x-linked-size"):
+        v = response.headers.get(h)
+        if v:
+            headers[h] = v
+    headers["location"] = absolute_location
+    # Presigned redirects expire — never let an intermediary cache a
+    # response whose target URL has a baked-in deadline.
+    headers["cache-control"] = "no-store"
+    strip_xet_response_headers(headers)
+    headers.update(add_source_headers(response, source["name"], source["url"]))
+    return Response(
+        status_code=response.status_code,
+        content=b"",
+        headers=headers,
+    )
+
 
 logger = get_logger("FALLBACK_OPS")
 
@@ -356,6 +452,7 @@ async def try_fallback_resolve(
     user_tokens: dict[str, str] | None = None,
     method: str = "GET",
     user=None,
+    client_headers: dict[str, str] | None = None,
 ) -> Optional[Response]:
     """Try to resolve file from fallback sources.
 
@@ -370,6 +467,12 @@ async def try_fallback_resolve(
         user: Authenticated user (or None for anonymous). Threaded
             through to the cache key as ``user_id`` for strict
             per-user binding isolation (#79).
+        client_headers: Client request headers to forward upstream.
+            Filtered through ``_filter_client_headers`` before any
+            outbound use, so callers may safely pass the raw
+            ``request.headers`` mapping — only Range / If-* survive
+            (Authorization / Cookie / Proxy-Authorization /
+            Accept-Encoding are dropped here).
 
     Returns:
         Response (redirect for GET, response with headers for HEAD) or None if not found
@@ -383,6 +486,10 @@ async def try_fallback_resolve(
 
     user_id = _resolve_user_id(user)
     tokens_hash = compute_tokens_hash(user_tokens)
+
+    # Defense-in-depth: drop everything that isn't on the resolve
+    # whitelist before it gets anywhere near the upstream chain.
+    safe_client_headers = _filter_client_headers(client_headers)
 
     # Construct KohakuHub path
     kohaku_path = f"/{repo_type}s/{namespace}/{name}/resolve/{revision}/{path}"
@@ -407,6 +514,7 @@ async def try_fallback_resolve(
             attempts,
             cache,
             gens,
+            client_headers=safe_client_headers,
         )
 
     return await _run_cached_then_chain(
@@ -433,6 +541,8 @@ async def _resolve_one_source(
     attempts: list[dict],
     cache,
     gens: tuple[int, int, int],
+    *,
+    client_headers: dict[str, str] | None = None,
 ) -> Optional[Response]:
     """Run a resolve probe (HEAD, then GET if method=GET) against one source.
 
@@ -544,10 +654,25 @@ async def _resolve_one_source(
     # the user gets. Falling through to another source here is the
     # cross-source mixing bug #75 fixes (HEAD-200 at A, GET-502 at A,
     # then sneak over to B's same-named-but-different repo).
+    #
+    # Plan A invariants enforced here:
+    #   • ``follow_redirects=False`` — never let httpx chase an upstream
+    #     30x into a CDN body fetch. A 1.5 GB safetensors must not pass
+    #     through the backend; the redirect Location is what we hand
+    #     back to the client (mirrors local ``resolve_file_get`` 302).
+    #   • ``headers=client_headers`` — forward the caller's whitelisted
+    #     Range / If-* headers so partial-content semantics survive.
+    #     The whitelist (Range, If-Match, If-None-Match,
+    #     If-Modified-Since, If-Unmodified-Since, If-Range) is built by
+    #     the ``with_repo_fallback`` decorator; Authorization / Cookie
+    #     are filtered there and never reach this call.
     get_t0 = time.monotonic()
     try:
         get_response = await client.get(
-            kohaku_path, repo_type, follow_redirects=True
+            kohaku_path,
+            repo_type,
+            follow_redirects=False,
+            headers=client_headers or None,
         )
     except httpx.TimeoutException as e:
         get_dt_ms = int((time.monotonic() - get_t0) * 1000)
@@ -594,6 +719,18 @@ async def _resolve_one_source(
         decision=classify_upstream(get_response),
         duration_ms=get_dt_ms,
     )
+
+    # Plan A primary path: 30x → forward Location to the client; the
+    # CDN/presigned target is the byte source, not this backend.
+    if (
+        300 <= get_response.status_code < 400
+        and get_response.headers.get("location")
+    ):
+        logger.info(
+            f"GET {get_response.status_code} → redirect-passthrough at "
+            f"{source['name']} (Location forwarded to client; no body buffer)"
+        )
+        return _propagate_upstream_redirect(get_response, source)
 
     if get_response.status_code == 200:
         # Proxy the content with original headers, stripping the

--- a/test/kohakuhub/api/fallback/test_decorators.py
+++ b/test/kohakuhub/api/fallback/test_decorators.py
@@ -141,16 +141,16 @@ async def test_with_repo_fallback_uses_resolve_operation_after_http_404(monkeypa
     hops = _decode_trace_header(result)
     assert any(h.get("decision") == "LOCAL_MISS" for h in hops)
     assert merged_inputs == [("owner-user", {"https://hf.local": "header-token"})]
-    assert resolve_calls == [
-        (
-            ("dataset", "owner", "demo", "main", "config.json"),
-            {
-                "user_tokens": {"https://hf.local": "token"},
-                "method": "HEAD",
-                "user": "owner-user",
-            },
-        )
-    ]
+    assert len(resolve_calls) == 1
+    args, kwargs = resolve_calls[0]
+    assert args == ("dataset", "owner", "demo", "main", "config.json")
+    assert kwargs["user_tokens"] == {"https://hf.local": "token"}
+    assert kwargs["method"] == "HEAD"
+    assert kwargs["user"] == "owner-user"
+    # Plan A: the decorator now also threads client_headers (raw
+    # request.headers) through; ``try_fallback_resolve`` is responsible
+    # for whitelisting Range/If-* and dropping everything else.
+    assert "client_headers" in kwargs
 
 
 @pytest.mark.asyncio

--- a/test/kohakuhub/api/fallback/test_operations.py
+++ b/test/kohakuhub/api/fallback/test_operations.py
@@ -2991,3 +2991,250 @@ async def test_lock_supervisor_releases_lock_when_attempt_fn_wedges(monkeypatch)
             f"response: {r!r}. The supervisor should surface a clean "
             f"aggregate failure, not propagate the cancellation."
         )
+
+
+# ---------------------------------------------------------------------------
+# Plan A: resolve GET as redirect-passthrough.
+#
+# Background: before this change, ``_resolve_one_source`` GET path issued
+# ``client.get(..., follow_redirects=True)`` and proxied the upstream body
+# through ``response.content`` — meaning a 1.5 GB safetensors download was
+# fully buffered into backend RAM before any byte reached the client, AND
+# the client's ``Range`` / ``If-*`` headers were silently dropped because
+# the call site forwarded no headers. Two failure modes:
+#
+#   (1) OOM / DoS — n concurrent large-file fetches × full file size
+#   (2) ~20 s 502 — supervisor ``wait_for`` cancels the runaway download
+#       (because the per-source httpx ``read`` timeout is an idle timeout,
+#       not a total-time timeout, so it never trips while the proxy keeps
+#       trickling bytes)
+#
+# Plan A: never proxy bytes for GET. ``follow_redirects=False`` upstream;
+# if upstream replies with 30x → forward ``Location`` to the client (same
+# shape as the local presigned-URL redirect in ``api/files.py``); if
+# upstream replies 200 inline (small files like ``config.json``) the
+# existing ``.content`` path is kept.
+# ---------------------------------------------------------------------------
+
+
+@pytest.mark.asyncio
+async def test_try_fallback_resolve_get_redirect_passthrough(monkeypatch):
+    """Bound source GET → 302 to a CDN: forward the redirect to the client
+    verbatim, do not chase ``Location`` from inside httpx, do not buffer
+    any redirect-target body."""
+    monkeypatch.setattr(fallback_ops, "get_cache", DummyCache)
+    monkeypatch.setattr(
+        fallback_ops,
+        "get_enabled_sources",
+        lambda namespace, user_tokens=None: [
+            {"url": "https://hf.local", "name": "HF", "source_type": "huggingface"},
+        ],
+    )
+    path = "/models/owner/demo/resolve/main/weights.safetensors"
+    FakeFallbackClient.queue("https://hf.local", "HEAD", path, _content_response(200))
+    cdn_location = "https://cas-bridge.xethub.hf.co/object?sig=xyz"
+    FakeFallbackClient.queue(
+        "https://hf.local",
+        "GET",
+        path,
+        _content_response(
+            302,
+            headers={
+                "location": cdn_location,
+                "etag": '"deadbeef"',
+                "x-repo-commit": "abc123",
+                "x-linked-size": "1519984962",
+            },
+        ),
+    )
+
+    response = await fallback_ops.try_fallback_resolve(
+        "model", "owner", "demo", "main", "weights.safetensors", method="GET",
+    )
+
+    assert response.status_code == 302
+    assert response.headers["location"] == cdn_location
+    assert response.headers["X-Source"] == "HF"
+    # Metadata HF clients read off the redirect must survive.
+    assert response.headers["x-repo-commit"] == "abc123"
+    assert response.headers["x-linked-size"] == "1519984962"
+    # ``follow_redirects=False`` was passed so httpx did not chase Location
+    # inside the client. No second GET against any source either.
+    get_calls = [c for c in FakeFallbackClient.calls if c[1] == "GET"]
+    assert len(get_calls) == 1, get_calls
+    assert get_calls[0][3].get("follow_redirects") is False
+
+
+@pytest.mark.asyncio
+async def test_try_fallback_resolve_get_forwards_range_header(monkeypatch):
+    """Client's ``Range`` / ``If-*`` request headers must reach the upstream
+    so a partial-content client request stays partial. Before Plan A the
+    ``FallbackClient.get`` call site never passed any headers — Range was
+    silently dropped and HF returned the full body."""
+    monkeypatch.setattr(fallback_ops, "get_cache", DummyCache)
+    monkeypatch.setattr(
+        fallback_ops,
+        "get_enabled_sources",
+        lambda namespace, user_tokens=None: [
+            {"url": "https://hf.local", "name": "HF", "source_type": "huggingface"},
+        ],
+    )
+    path = "/models/owner/demo/resolve/main/weights.safetensors"
+    FakeFallbackClient.queue("https://hf.local", "HEAD", path, _content_response(200))
+    FakeFallbackClient.queue(
+        "https://hf.local",
+        "GET",
+        path,
+        _content_response(302, headers={"location": "https://cdn/obj"}),
+    )
+
+    await fallback_ops.try_fallback_resolve(
+        "model", "owner", "demo", "main", "weights.safetensors",
+        method="GET",
+        client_headers={
+            "Range": "bytes=0-100000",
+            "If-None-Match": '"abc"',
+            # Not on the whitelist — must NOT propagate. Forwarding
+            # ``Accept-Encoding`` would also re-engage httpx's auto-decompression
+            # which is incompatible with redirect-passthrough.
+            "Accept-Encoding": "gzip",
+            "User-Agent": "kohakuhub/test",
+        },
+    )
+
+    get_calls = [c for c in FakeFallbackClient.calls if c[1] == "GET"]
+    assert len(get_calls) == 1
+    fwd = {k.lower(): v for k, v in (get_calls[0][3].get("headers") or {}).items()}
+    assert fwd.get("range") == "bytes=0-100000"
+    assert fwd.get("if-none-match") == '"abc"'
+    assert "accept-encoding" not in fwd
+    assert "user-agent" not in fwd
+
+
+@pytest.mark.asyncio
+async def test_try_fallback_resolve_get_never_forwards_authorization_or_cookie(monkeypatch):
+    """Safety: client's ``Authorization`` / ``Cookie`` /
+    ``Proxy-Authorization`` MUST NEVER reach the upstream. The only
+    credential that may travel upstream is the admin-configured source
+    token, attached by ``FallbackClient`` itself."""
+    monkeypatch.setattr(fallback_ops, "get_cache", DummyCache)
+    monkeypatch.setattr(
+        fallback_ops,
+        "get_enabled_sources",
+        lambda namespace, user_tokens=None: [
+            {"url": "https://hf.local", "name": "HF", "source_type": "huggingface"},
+        ],
+    )
+    path = "/models/owner/demo/resolve/main/weights.safetensors"
+    FakeFallbackClient.queue("https://hf.local", "HEAD", path, _content_response(200))
+    FakeFallbackClient.queue(
+        "https://hf.local",
+        "GET",
+        path,
+        _content_response(302, headers={"location": "https://cdn/obj"}),
+    )
+
+    await fallback_ops.try_fallback_resolve(
+        "model", "owner", "demo", "main", "weights.safetensors",
+        method="GET",
+        client_headers={
+            "Range": "bytes=0-100000",
+            "Authorization": "Bearer user-secret",
+            "Cookie": "session_id=secret-cookie",
+            "Proxy-Authorization": "Basic Zm9vOmJhcg==",
+        },
+    )
+
+    get_calls = [c for c in FakeFallbackClient.calls if c[1] == "GET"]
+    fwd = {k.lower(): v for k, v in (get_calls[0][3].get("headers") or {}).items()}
+    assert fwd.get("range") == "bytes=0-100000"
+    assert "authorization" not in fwd
+    assert "cookie" not in fwd
+    assert "proxy-authorization" not in fwd
+
+
+@pytest.mark.asyncio
+async def test_try_fallback_resolve_get_redirect_rewrites_relative_location(monkeypatch):
+    """HF's non-LFS resolve-cache pattern returns a 307 with a *relative*
+    Location like ``/api/resolve-cache/...``. Plan A must rewrite that
+    to absolute against the upstream URL so the client follows the
+    redirect on the upstream itself — NOT back to KohakuHub which
+    does not serve ``/api/resolve-cache/``."""
+    monkeypatch.setattr(fallback_ops, "get_cache", DummyCache)
+    monkeypatch.setattr(
+        fallback_ops,
+        "get_enabled_sources",
+        lambda namespace, user_tokens=None: [
+            {"url": "https://hf.local", "name": "HF", "source_type": "huggingface"},
+        ],
+    )
+    path = "/models/owner/demo/resolve/main/pattern_a.txt"
+    FakeFallbackClient.queue("https://hf.local", "HEAD", path, _content_response(200))
+    FakeFallbackClient.queue(
+        "https://hf.local",
+        "GET",
+        path,
+        _content_response(
+            307,
+            headers={
+                "location": "/api/resolve-cache/models/owner/demo/sha/pattern_a.txt",
+                "x-repo-commit": "abc123",
+            },
+            url="https://hf.local/owner/demo/resolve/main/pattern_a.txt",
+        ),
+    )
+
+    response = await fallback_ops.try_fallback_resolve(
+        "model", "owner", "demo", "main", "pattern_a.txt", method="GET",
+    )
+
+    assert response.status_code == 307
+    # urljoin against the upstream URL gives us the upstream-absolute
+    # path; the client now follows on the upstream rather than bouncing
+    # the relative path back to our own backend.
+    assert (
+        response.headers["location"]
+        == "https://hf.local/api/resolve-cache/models/owner/demo/sha/pattern_a.txt"
+    )
+    assert response.headers["x-repo-commit"] == "abc123"
+
+
+@pytest.mark.asyncio
+async def test_try_fallback_resolve_get_inline_200_still_returns_body(monkeypatch):
+    """Plan A only changes the redirect path. A small inline 200 (e.g.
+    ``config.json``) still passes the body through verbatim with the
+    historical ``Content-Encoding`` / ``Content-Length`` /
+    ``Transfer-Encoding`` strip applied."""
+    monkeypatch.setattr(fallback_ops, "get_cache", DummyCache)
+    monkeypatch.setattr(
+        fallback_ops,
+        "get_enabled_sources",
+        lambda namespace, user_tokens=None: [
+            {"url": "https://hf.local", "name": "HF", "source_type": "huggingface"},
+        ],
+    )
+    path = "/models/owner/demo/resolve/main/config.json"
+    FakeFallbackClient.queue("https://hf.local", "HEAD", path, _content_response(200))
+    FakeFallbackClient.queue(
+        "https://hf.local",
+        "GET",
+        path,
+        _content_response(
+            200,
+            b'{"hello":"world"}',
+            headers={"content-type": "application/json"},
+        ),
+    )
+
+    response = await fallback_ops.try_fallback_resolve(
+        "model", "owner", "demo", "main", "config.json", method="GET",
+    )
+
+    assert response.status_code == 200
+    assert response.body == b'{"hello":"world"}'
+    assert response.headers["X-Source"] == "HF"
+    # The GET upstream must still have been issued with redirect chasing
+    # disabled — the same ``follow_redirects=False`` invariant applies even
+    # when the response happens to be a direct 200.
+    get_calls = [c for c in FakeFallbackClient.calls if c[1] == "GET"]
+    assert get_calls[0][3].get("follow_redirects") is False

--- a/test/kohakuhub/api/fallback/test_operations.py
+++ b/test/kohakuhub/api/fallback/test_operations.py
@@ -3238,3 +3238,141 @@ async def test_try_fallback_resolve_get_inline_200_still_returns_body(monkeypatc
     # when the response happens to be a direct 200.
     get_calls = [c for c in FakeFallbackClient.calls if c[1] == "GET"]
     assert get_calls[0][3].get("follow_redirects") is False
+
+
+@pytest.mark.asyncio
+async def test_try_fallback_resolve_get_redirect_drops_xet_headers_from_upstream(monkeypatch):
+    """Plan A redirect-passthrough must NOT leak ``x-xet-*`` headers
+    upstream might attach. The output is built from an explicit
+    six-key whitelist (``etag`` / ``x-repo-commit`` / ``x-linked-etag``
+    / ``x-linked-size`` / ``location`` / ``cache-control``), so any
+    Xet-shaped key on the upstream response must not survive — even
+    if the (defensive) ``strip_xet_response_headers`` call is later
+    removed as redundant. This test locks the contract at the
+    response surface so that refactor stays safe."""
+    monkeypatch.setattr(fallback_ops, "get_cache", DummyCache)
+    monkeypatch.setattr(
+        fallback_ops,
+        "get_enabled_sources",
+        lambda namespace, user_tokens=None: [
+            {"url": "https://hf.local", "name": "HF", "source_type": "huggingface"},
+        ],
+    )
+    path = "/models/owner/demo/resolve/main/weights.safetensors"
+    FakeFallbackClient.queue("https://hf.local", "HEAD", path, _content_response(200))
+    FakeFallbackClient.queue(
+        "https://hf.local",
+        "GET",
+        path,
+        _content_response(
+            302,
+            headers={
+                "location": "https://cas-bridge.xethub.hf.co/object?sig=xyz",
+                "etag": '"deadbeef"',
+                "x-repo-commit": "abc123",
+                "x-linked-size": "1519984962",
+                # Hostile / accidental upstream headers we must drop.
+                "x-xet-cas-url": "https://cas.xethub.hf.co",
+                "x-xet-hash": "shardhash",
+                "x-xet-cas-uid": "public",
+                "x-xet-something-future": "still-dropped",
+            },
+        ),
+    )
+
+    response = await fallback_ops.try_fallback_resolve(
+        "model", "owner", "demo", "main", "weights.safetensors", method="GET",
+    )
+
+    assert response.status_code == 302
+    assert response.headers["location"] == "https://cas-bridge.xethub.hf.co/object?sig=xyz"
+    # Whitelist survivors.
+    assert response.headers["etag"] == '"deadbeef"'
+    assert response.headers["x-repo-commit"] == "abc123"
+    assert response.headers["x-linked-size"] == "1519984962"
+    # Xet hints must NOT be present — this enforces the contract at
+    # the OUTPUT regardless of how the curation logic is implemented.
+    lower_keys = {k.lower() for k in response.headers.keys()}
+    assert not any(k.startswith("x-xet-") for k in lower_keys), (
+        f"x-xet-* headers leaked into redirect response: "
+        f"{[k for k in lower_keys if k.startswith('x-xet-')]}"
+    )
+
+
+@pytest.mark.asyncio
+async def test_try_fallback_resolve_head_does_not_forward_client_headers(monkeypatch):
+    """HEAD path is intentionally asymmetric with GET on header
+    forwarding: ``client.head`` is invoked WITHOUT ``client_headers``.
+    Two reasons (kept here as the regression-guard for the next
+    refactor that's tempted to "make HEAD symmetric"):
+
+      1) ``huggingface_hub`` HEAD-on-resolve never carries Range — the
+         metadata probe is positional, partial-content semantics are a
+         GET-only concern.
+      2) ``apply_resolve_head_postprocess`` fires its own follow-HEAD
+         with ``Accept-Encoding: identity`` deliberately set so httpx
+         doesn't auto-decompress and silently strip Content-Length
+         (PR #21). Forwarding a client-supplied ``Accept-Encoding:
+         gzip`` upstream would re-engage that bug.
+
+    If a future PR wants ``If-None-Match`` 304 short-circuit on HEAD,
+    the right fix is a NARROWER whitelist forwarded to the binding
+    HEAD probe ONLY — never to the follow-HEAD inside
+    apply_resolve_head_postprocess. This test fails loudly if that
+    boundary is crossed (any client header reaches the binding
+    HEAD probe)."""
+    monkeypatch.setattr(fallback_ops, "get_cache", DummyCache)
+    monkeypatch.setattr(
+        fallback_ops,
+        "get_enabled_sources",
+        lambda namespace, user_tokens=None: [
+            {"url": "https://hf.local", "name": "HF", "source_type": "huggingface"},
+        ],
+    )
+    path = "/models/owner/demo/resolve/main/weights.safetensors"
+    FakeFallbackClient.queue(
+        "https://hf.local",
+        "HEAD",
+        path,
+        _content_response(
+            302,
+            headers={
+                "location": "https://cas-bridge.xethub.hf.co/object?sig=xyz",
+                "x-linked-size": "1519984962",
+                "x-linked-etag": '"deadbeef"',
+                "x-repo-commit": "abc123",
+            },
+        ),
+    )
+
+    await fallback_ops.try_fallback_resolve(
+        "model", "owner", "demo", "main", "weights.safetensors",
+        method="HEAD",
+        # Caller passes everything — Range, conditional, even
+        # Authorization (which the operations-layer filter normally
+        # drops anyway). The HEAD probe must receive none of it.
+        client_headers={
+            "Range": "bytes=0-100000",
+            "If-None-Match": '"abc"',
+            "Authorization": "Bearer secret",
+        },
+    )
+
+    head_calls = [c for c in FakeFallbackClient.calls if c[1] == "HEAD"]
+    assert len(head_calls) == 1
+    head_kwargs = head_calls[0][3]
+    # Two equivalent guards:
+    # (a) ``headers`` kwarg either absent or empty/None — the call
+    #     site does ``await client.head(kohaku_path, repo_type)`` with
+    #     no extra args.
+    fwd_headers = head_kwargs.get("headers") or {}
+    assert not fwd_headers, (
+        f"HEAD probe leaked client headers: {dict(fwd_headers)}. "
+        f"See PR #21 — Accept-Encoding propagation breaks Content-Length."
+    )
+    # (b) Specifically none of the whitelist or auth headers may
+    #     appear, even if a future implementation refactors how the
+    #     ``headers`` kwarg is shaped.
+    fwd_lower = {k.lower(): v for k, v in fwd_headers.items()}
+    for forbidden in ("range", "if-none-match", "authorization"):
+        assert forbidden not in fwd_lower

